ToolsPanel: Allow SlotFill injection of panel items

aaronrobertshaw opened the pull request:
Related: 
- https://github.com/WordPress/gutenberg/pull/34157
- https://github.com/WordPress/gutenberg/pull/34065
- https://github.com/WordPress/gutenberg/pull/34530
- https://github.com/WordPress/gutenberg/pull/34063
- https://github.com/WordPress/gutenberg/pull/34069

## Description

This updates the `ToolsPanel` component to support injection of panel items via a SlotFill.  The changes to the `ToolsPanel` here have been separated from https://github.com/WordPress/gutenberg/pull/34157.

#### Changes include:
- Exporting `ToolsPanelContext` so it can be passed via slots' `fillProps`
- Merges the `label` and `header` props for the `ToolsPanel`
- Updates readmes
- Fixes issues that arise when panel items are added via a shared slot 
    - e.g. items possibly still being in the slot's fills when changing block selections and the panel effectively changing
- Add ability to supply filters to the `resetAll` callback allowing injected panel item controls to reset whatever they need

## How has this been tested?

- Ensured tests are passing, including updated test for merged props:
    - `npm run test-unit packages/components/src/tools-panel/test`
- Manually through the storybook: 
    - `npm run storybook:dev`
    - http://localhost:50240/?path=/story/components-experimental-toolspanel--default
- Manually within the block editor, adding a group block and its "dimensions" panel

## Screenshots <!-- if applicable -->

![ToolsPanelPostSlotFillUpdates](https://user-images.githubusercontent.com/60436221/132454586-70965a98-b91f-479b-8bcb-19a6dc66438c.gif)


## Types of changes
Enhancement. Bug fix.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

aaronrobertshaw committed:
Update ToolsPanel to allow SlotFill injection of panel items

gziolo commented:
I see changes in the public API but it seems to be fine `ToolsPanel` and `ToolsPanelItem` are still considered experimental:

https://github.com/WordPress/gutenberg/blob/e3591aa86bfc4fe02da616ef50e722504ba8b3d8/packages/components/src/index.js#L141-L144

ciampo reviewed (commented):
Thank you for moving changes to the `ToolsPanel` to a separate PR, it made this review much easier!

Code changes in this PR LGTM :) Thank you also for adding more props to the docs, and for updating unit tests!

I was playing around with the component in Storybook, and probably came across a bug: deselecting a component doesn't seem to reset its value _after_ clicking on the "Reset All" button:

https://user-images.githubusercontent.com/1083581/132499359-00cd3380-c8d6-4fac-936c-3aec8c80710e.mp4

Also, do you think it'd be valuable to add a Storybook example showing how to use the `panelId` props, especially in the context of injection of panel items via a SlotFill?

Although I'm ok with not adding it if you think it doesn't make much sense / bring much value.

gziolo commented on packages/components/src/tools-panel/tools-panel/README.md:
```diff
@@ -73,19 +69,22 @@ export function DimensionPanel( props ) {
 - Required: Yes
 
-### `resetAll`: `function`
+### `panelId`: `function`
```
> Also, do you think it'd be valuable to add a Storybook example showing how to use the panelId props, especially in the context of injection of panel items via a SlotFill?

Continuing on the comment from @ciampo, it feels like this prop needs a better description that is disconnected from the SlotFill pattern. `@wordpress/components` is a set of general purpose components so it feels like this part is tightly couple with the higher-level abstraction that doesn't fit here.

aaronrobertshaw replied:
I've had another go at [describing this prop](https://github.com/WordPress/gutenberg/pull/34632/commits/ee6b6821b9a5960f56ec3c8e38f513c31600c58a) free of mentioning the SlotFill pattern. Let me know if that misses the mark and needs some further rewording.

gziolo replied:
Thank you. I like more general phrasing more 👍🏻

aaronrobertshaw committed:
Fix clearing of isResetting ref

aaronrobertshaw committed:
Add SlotFill demo for ToolsPanel story

aaronrobertshaw committed:
Improve ToolsPanel readmes

aaronrobertshaw commented:
> I was playing around with the component in Storybook, and probably came across a bug: deselecting a component doesn't seem to reset its value after clicking on the "Reset All" button:

Thanks for catching that @ciampo! 

When addressing the last round of feedback, I messed up the clearing of the `isResetting` ref which ultimately caused the panel items to think the panel was still undertaking a "reset all" and therefore skip calling the `onDeselect` callbacks.

I've added another unit test for this along with a fix. I now get the following behaviour:

https://user-images.githubusercontent.com/60436221/132624756-cd71b628-b79b-4853-94a8-89327d98de8a.mp4

> Also, do you think it'd be valuable to add a Storybook example showing how to use the panelId props, especially in the context of injection of panel items via a SlotFill?

I've added another storybook example that creates its own SlotFill and uses that to illustrate injecting items into the panel. Hopefully, that makes sense and is useful. I'm not tied to it so if anyone has any suggestions on improving it or even removing it, that's fine by me.

The rewording of the docs and some typo fixes have been made as well.

gziolo commented on packages/components/src/tools-panel/stories/index.js:
```diff
@@ -76,7 +77,103 @@ export const _default = () => {
 	);
 };
 
+const { Fill, Slot } = createSlotFill( 'ToolsPanelSlot' );
+
+const ToolsPanelItems = ( { children } ) => {
+	return <Fill>{ children }</Fill>;
+};
```
Would it work as follows?
```suggestion
const { Fill: ToolsPanelItems, Slot } = createSlotFill( 'ToolsPanelSlot' );
```

aaronrobertshaw replied:
It does work although I get a warning about the `Slot` property not existing on type. Which can be avoided by simply using the `Slot` direct from `createSlotFill`. I was trying to stay consistent with how most other SlotFills are handled in other packages.

I'll update this so the code is as simple as possible.

aaronrobertshaw replied:
✅ Done. Thanks for the suggestion.

aaronrobertshaw committed:
Simplify handling of SlotFill ToolsPanel example

gziolo commented on packages/components/src/tools-panel/stories/index.js:
```diff
@@ -79,7 +77,97 @@ export const _default = () => {
+
+	return (
+		<SlotFillProvider>
+			<ToolsPanelItems>
```
This example and the documentation efforts made me think whether `panelId` should be part of the component. Should we create an opinionated Fill and Slot and keep them in the block editor instead?

The difference would be that `ToolsPanelItem` from the `@wordpress/block-editor` package would be a fill that handles all the logic related to `panelId` and renders `ToolsPanelItem`.  There would be also a slot that would pass down `panelId` to fills through `fillProps` and would be wrapped with `ToolPanel` component. I don't know if I'm doing a good job explaining it. 

So the goal would be to achieve the following:

```js
const { Fill, Slot } = createSlotFill( 'ToolsPanelSlot' );

function shouldRender( panelId ) {
    // checks somehow whether to render the panel based on the value passed from slot using `fillProps`
}

const ToolsPanelItemFill = ( { panelId, ...props } ) => {
    return (
        <Fill>
            { shouldRender( panelId ) && <ToolsPanelItem { ...props } /> }
        </Fill>
    );
};

const ToolsPanelSlot = ( { panelId, ...props } ) => {
    return (
        <ToolsPanel { ...props }>
            <Slot fillProps={ panelId } />
        </ToolsPanel>
    );
}; 
```

It needs more code, but I hope it helps to sketch the idea.

aaronrobertshaw replied:
That's an interesting idea. I don't think I fully understand the implications yet but will explore this tomorrow.

It makes sense in terms of being able to remove the need for `panelId` on the `ToolsPanel` itself. At this stage though, I don't have much of an idea how the individual `ToolsPanelItem`s would be working.

gziolo replied:
You probably can play with the Storybook story first and see if you can make it work this way. As noted in another place, given that both updated components are marked as experimental don't treat my comment as a blocker.

gziolo reviewed (approved):
Feel free to merge when @ciampo approves the recent changes 👍🏻

ciampo commented on packages/components/src/tools-panel/tools-panel/README.md:
```diff
@@ -73,19 +69,22 @@ export function DimensionPanel( props ) {
 ### `label`: `string`
 
-The label for the panel's dropdown menu.
+Text to be displayed within the panel's header and as the aria-label for the
```
```suggestion
Text to be displayed within the panel's header and as the `aria-label` for the
```

ciampo reviewed (approved):
Thank you @aaronrobertshaw for addressing all the feedback!

> When addressing the last round of feedback, I messed up the clearing of the isResetting ref which ultimately caused the panel items to think the panel was still undertaking a "reset all" and therefore skip calling the onDeselect callbacks.

This explanation makes total sense and explains the bug — I can confirm that the component now behaves as expected!

> I've added another storybook example that creates its own SlotFill and uses that to illustrate injecting items into the panel.

I agree with @gziolo's [point](https://github.com/WordPress/gutenberg/pull/34632#discussion_r705068142) about potentially extracting `panelId` (and slot fill logic) to another part of the repo, but that's definitely not a blocker and should be investigated/tackled separately.

Also, special thanks for improving unit tests with every PR!

aaronrobertshaw committed:
Improve label prop description

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>
